### ISSUE TITLE
Jax to numpy process takes too long time on TPU v4-8 for FlaxStableDiffusionXLPipeline

### ISSUE THREAD

**Issue by gokerguner** [huggingface/diffusers#5277]
Hello, I'm trying to FlaxStableDiffusionXLPipeline implementation on TPU v4-8. And I measure the time of each step.

When I give the prompts in groups of 8, the 8-image generation time takes 0.02 seconds (after compiling).

Although this time seems very good, converting the `images` variable from jax to numpy takes around 4.68 seconds and this destroys TPU efficiency.

Is it possible to make this change in a shorter time?

The first part is nearly same to example:
```
import numpy as np
import jax
import jax.numpy as jnp
import random
from pathlib import Path
from jax import pmap
from flax.jax_utils import replicate
from flax.training.common_utils import shard
from PIL import Image
import time
import prompts
from diffusers import FlaxStableDiffusionXLPipeline, FlaxDPMSolverMultistepScheduler, FlaxEulerDiscreteScheduler, FlaxStableDiffusionPipeline
from typing import List, Optional, Tuple, Union

num_devices = jax.device_count()
device_type = jax.devices()[0].device_kind
device = jax.devices()[0]
print(f"Found {num_devices} JAX devices of type {device_type}.")
assert (
    "TPU" in device_type
), "Available device is not a TPU, please select TPU from Edit > Notebook settings > Hardware accelerator"

dtype = jnp.float32

pipeline, params = FlaxStableDiffusionXLPipeline.from_pretrained(
    "pcuenq/stable-diffusion-xl-base-1.0-flax",
    dtype=dtype
)

scheduler_id = "stabilityai/stable-diffusion-xl-base-1.0"
scheduler, state = FlaxDPMSolverMultistepScheduler.from_pretrained(scheduler_id, subfolder="scheduler")
pipeline.scheduler = scheduler

params = jax.tree_util.tree_map(lambda x: x.astype(dtype), params)
params["scheduler"] = state

def create_key(seed=0):
    return jax.random.PRNGKey(seed)

def chunk_list(input_list, chunk_size):
    return [input_list[i:i+chunk_size] for i in range(0, len(input_list), chunk_size)]

chunks = chunk_list(prompts.all_prompts, 8)
rng = create_key()
rng = jax.random.split(rng, jax.device_count())
p_params = replicate(params)
number = 0

batch_size, height, width, ch = 8, 128, 128, 4

np.random.seed(0)
latents_shape = (batch_size, ch, height, width)
latents = jnp.array(np.random.randn(*latents_shape).astype(np.float32),dtype=jnp.float32)
latents = latents * 1.0
latents = shard(latents)

prompt_ids = pipeline.prepare_inputs(chunks[0])
prompt_ids = shard(prompt_ids)
images = pipeline(prompt_ids, p_params, rng, jit=True, latents=latents, num_inference_steps=20)[0] # First compile out of the for loop
```

Then I start to give my prompts. I have 72 prompts. I give these to TPU in batches of 8, for a total of 9 batches.

```
init_start = time.time()
batch_num = 1
for chunk in chunks:
    start = time.time()
    prompt_ids = pipeline.prepare_inputs(chunk)
    prompt_ids = shard(prompt_ids)
    end_shard = time.time()
    print(batch_num, " inputs sharded in: "+str(round(end_shard - start,2))) # It takes 0.01 sec
    try:
        images = pipeline(prompt_ids, p_params, rng, jit=True, latents=latents, num_inference_steps=20)[0]
    except ValueError as e:
        print(str(e))
        continue
    end = time.time()
    print(batch_num, " finished, elapsed time(second) for 8-image batch: " + str(round(end - start,2))) # It takes 0.02
    start_npc = time.time()
    images = np.array(images)
    end_npc = time.time()
    print(batch_num, " jax to numpy " + str(round(end_npc - start_npc,2))) # IT TAKES 4.68, HERE IS THE PROBLEM
    start_reshape = time.time()
    images = images.reshape((images.shape[0] * images.shape[1],) + images.shape[-3:])
    end_reshape = time.time()
    print(batch_num, " reshaped: "+ str(round(end_reshape - start_reshape,2))) # It takes nearly 0.0 sec
    start_nppil = time.time()
    images = pipeline.numpy_to_pil(np.array(images))
    end_nppil = time.time()
    print(batch_num, " numpy to pil " + str(round(end_nppil - start_nppil,2))) # It takes 0.06
    end_process = time.time()
    print(batch_num, " images post-processes in: "+str(round(end_process - start,2))) # Total post process takes 4.77 sec cause of jax to numpy challenge
    for img in images:
        name = "no_grid_outputs_sdxl/output" + str(number) + ".png"
        number += 1
        img.save(name)
    end_batch = time.time()
    print(batch_num, " images saved in: "+ str(round(end_batch - start,2))) # Total process takes nearly 8.11 sec
    print("*************")
    batch_num += 1

init_end = time.time()
print("Elapsed total time(second): " + str(round(init_end - init_start,2)))
```

I wrote the times as comments to the right of the relevant print lines. However, I leave a sample log below.

```
inputs sharded in: 0.01
finished, elapsed time(second) for 8-image batch: 0.03
jax to numpy 4.68
reshaped: 0.0
numpy to pil 0.06
images post-processes in: 4.76
images saved in: 7.88
```

Post-process and save times are cumulative. I want to do jax to numpy process cause of saving the images. Final words;
- How can I reduce jax to numpy process time?
- Are there any other way to save images without convertion to numpy?

**Comment by patrickvonplaten** [comment]
> When I give the prompts in groups of 8, the 8-image generation time takes 0.02 seconds (after compiling).
This would be incredibly fast, I doubt a bit that something like this can be achieved tbh

cc @pcuenca as well

**Comment by pcuenca** [comment]
Hi @gokerguner! I don't think it's the numpy conversion what's taking time, I believe what you are seeing is because of the asynchronous dispatch of JAX operations. When you log your `finished` time, the operation is actually not finished. When you request the conversion to `numpy`, then the Python interpreter waits until generation completes and data becomes available.

You can try to insert the following line before your log to ensure that image generation has actually finished:

```python
images.block_until_ready()
```

There's more details about this behaviour here: https://jax.readthedocs.io/en/latest/async_dispatch.html

I'd say 4s for 8 images is also great :)

**Comment by gokerguner** [comment]
Yes, it really looks like that, it takes 4.67 sec with your suggestion :(. I have one more question about this: 

I was measuring the time with the same script in my Stable Diffusion experiments (it took approximately 1.1 seconds). Thinking that this measurement might be wrong, I tried to use the block_until_ready method there as well. I got an error saying “numpy.ndarray has no such method”. In this case, is the 1.1 seconds I measured for Stable Diffusion correct (for 8-batch)?
